pkg/spdk/engine.go

```diff
@@ -180,6 +156,38 @@ func getBdevNameForReplica(spdkClient *spdkclient.Client, localReplicaBdevMap ma
 	return nvmeBdevNameList[0], nil
 }
 
+func (e *Engine) handleFrontend(spdkClient *spdkclient.Client, superiorPortAllocator *util.Bitmap) error {
+	if e.Frontend != types.FrontendSPDKVoid && e.Frontend != types.FrontendSPDKTCPNvmf && e.Frontend != types.FrontendSPDKTCPBlockdev {
+		return fmt.Errorf("unknown frontend type %s", e.Frontend)
+	}
+
+	if e.Frontend == types.FrontendSPDKVoid {
+		e.log.Infof("No frontend specified, will not expose the volume %s", e.VolumeName)
+		return nil
+	}
+
+	nqn := helpertypes.GetNQN(e.Name)
+	port, _, err := superiorPortAllocator.AllocateRange(1)
+	if err != nil {
+		return err
+	}
+	if err := spdkClient.StartExposeBdev(nqn, e.Name, e.IP, strconv.Itoa(int(port))); err != nil {
+		return err
+	}
+	e.Port = port
+
+	if e.Frontend == types.FrontendSPDKTCPNvmf {
+		e.Endpoint = GetNvmfEndpoint(nqn, e.IP, e.Port)
```

Comment:
The format `fmt.Sprintf("%s://%s:%d", nqn, ip, port)` looks a bit weird. Is it correct?

Reply:
Do we need to follow the same pattern as frontend `iscsi`? For example `iscsi://10.42.0.111:3260/iqn.2019-10.io.longhorn:vol/1`.

Reply:
Yes, maybe `nvmf://<ip>:<port>/<nqn>`?